File: alphaforge/model/enums.py
# AlphaForge Enumerations
"""
Enumeration types for AlphaForge trading system.
"""

from enum import Enum, IntEnum


class OrderSide(IntEnum):
    """Order side enumeration."""
    NO_ORDER_SIDE = 0
    BUY = 1
    SELL = 2


class OrderType(IntEnum):
    """Order type enumeration."""
    NO_ORDER_TYPE = 0
    MARKET = 1
    LIMIT = 2
    STOP = 3
    STOP_LIMIT = 4
    MARKET_TO_LIMIT = 5
    MARKET_IF_TOUCHED = 6
    LIMIT_IF_TOUCHED = 7
    TRAILING_STOP_MARKET = 8
    TRAILING_STOP_LIMIT = 9


class OrderStatus(IntEnum):
    """Order status enumeration."""
    INITIALIZED = 0
    INVALID = 1
    DENIED = 2
    EMULATED = 3
    RELEASED = 4
    PENDING_UPDATE = 5
    PENDING_CANCEL = 6
    ACCEPTED = 7
    REJECTED = 8
    CANCELED = 9
    EXPIRED = 10
    TRIGGERED = 11
    PENDING_NEW = 12
    PARTIALLY_FILLED = 13
    FILLED = 14


class TimeInForce(IntEnum):
    """Time in force enumeration."""
    NO_TIME_IN_FORCE = 0
    DAY = 1
    GTC = 2  # Good Till Cancel
    IOC = 3  # Immediate Or Cancel
    FOK = 4  # Fill Or Kill
    GTD = 5  # Good Till Date
    AT_THE_OPEN = 6
    AT_THE_CLOSE = 7


class PositionSide(IntEnum):
    """Position side enumeration."""
    NO_POSITION_SIDE = 0
    FLAT = 1
    LONG = 2
    SHORT = 3


class LiquiditySide(IntEnum):
    """Liquidity side enumeration."""
    NO_LIQUIDITY_SIDE = 0
    MAKER = 1
    TAKER = 2


class OrderBookAction(IntEnum):
    """Order book action enumeration."""
    ADD = 1
    UPDATE = 2
    DELETE = 3
    CLEAR = 4


class MarketStatus(IntEnum):
    """Market status enumeration."""
    CLOSED = 0
    PRE_OPEN = 1
    OPEN = 2
    PAUSE = 3
    HALT = 4
    RESUME = 5
    PRE_CLOSE = 6


class InstrumentClass(IntEnum):
    """Instrument class enumeration."""
    SPOT = 1
    SWAP = 2
    FUTURE = 3
    FORWARD = 4
    CFD = 5
    BOND = 6
    OPTION = 7
    WARRANT = 8


class AssetClass(IntEnum):
    """Asset class enumeration."""
    FX = 1
    EQUITY = 2
    COMMODITY = 3
    DEBT = 4
    INDEX = 5
    CRYPTOCURRENCY = 6
    ALTERNATIVE = 7


class CurrencyType(IntEnum):
    """Currency type enumeration."""
    ISO = 1
    CRYPTO = 2
    COMMODITY_BACKED = 3


class AccountType(IntEnum):
    """Account type enumeration."""
    CASH = 1
    MARGIN = 2
    BETTING = 3


class OmsType(IntEnum):
    """Order Management System type."""
    UNSPECIFIED = 0
    NETTING = 1
    HEDGING = 2


class StrategyStatus(IntEnum):
    """Strategy status enumeration."""
    INITIALIZED = 0
    STARTING = 1
    RUNNING = 2
    STOPPING = 3
    STOPPED = 4
    DISPOSED = 5


class ComponentStatus(IntEnum):
    """Component status enumeration."""
    PRE_INITIALIZED = 0
    READY = 1
    STARTING = 2
    RUNNING = 3
    STOPPING = 4
    STOPPED = 5
    RESUMING = 6
    RESETTING = 7
    DISPOSING = 8
    DISPOSED = 9
    DEGRADED = 10
    FAULTED = 11


class LogLevel(IntEnum):
    """Logging level enumeration."""
    DEBUG = 10
    INFO = 20
    WARNING = 30
    ERROR = 40
    CRITICAL = 50


class AggressorSide(IntEnum):
    """Aggressor side for trades."""
    NO_AGGRESSOR = 0
    BUYER = 1
    SELLER = 2


class BookType(IntEnum):
    """Order book type enumeration."""
    L1_TBBO = 1  # Top of Book Best Bid/Offer
    L2_MBP = 2   # Market By Price
    L3_MBO = 3   # Market By Order


class PriceType(IntEnum):
    """Price type enumeration."""
    BID = 1
    ASK = 2
    MID = 3
    LAST = 4


class BarAggregation(IntEnum):
    """Bar aggregation method."""
    TICK = 1
    TICK_IMBALANCE = 2
    TICK_RUNS = 3
    VOLUME = 4
    VOLUME_IMBALANCE = 5
    VOLUME_RUNS = 6
    VALUE = 7
    VALUE_IMBALANCE = 8
    VALUE_RUNS = 9
    MILLISECOND = 10
    SECOND = 11
    MINUTE = 12
    HOUR = 13
    DAY = 14
    WEEK = 15
    MONTH = 16


class TriggerType(IntEnum):
    """Trigger type for conditional orders."""
    NO_TRIGGER = 0
    DEFAULT = 1
    BID_ASK = 2
    LAST = 3
    DOUBLE_LAST = 4
    DOUBLE_BIDASK = 5
    LAST_OR_BID_ASK = 6
    MID_POINT = 7
    MARK_PRICE = 8
    INDEX_PRICE = 9


class ContingencyType(IntEnum):
    """Contingency type for related orders."""
    NO_CONTINGENCY = 0
    OTO = 1  # One-Triggers-Other
    OCO = 2  # One-Cancels-Other
    OUO = 3  # One-Updates-Other


# String mappings for display
ORDER_SIDE_MAP = {
    OrderSide.NO_ORDER_SIDE: "NO_ORDER_SIDE",
    OrderSide.BUY: "BUY", 
    OrderSide.SELL: "SELL"
}

ORDER_TYPE_MAP = {
    OrderType.NO_ORDER_TYPE: "NO_ORDER_TYPE",
    OrderType.MARKET: "MARKET",
    OrderType.LIMIT: "LIMIT",
    OrderType.STOP: "STOP",
    OrderType.STOP_LIMIT: "STOP_LIMIT",
    OrderType.MARKET_TO_LIMIT: "MARKET_TO_LIMIT",
    OrderType.MARKET_IF_TOUCHED: "MARKET_IF_TOUCHED",
    OrderType.LIMIT_IF_TOUCHED: "LIMIT_IF_TOUCHED",
    OrderType.TRAILING_STOP_MARKET: "TRAILING_STOP_MARKET",
    OrderType.TRAILING_STOP_LIMIT: "TRAILING_STOP_LIMIT"
}

ORDER_STATUS_MAP = {
    OrderStatus.INITIALIZED: "INITIALIZED",
    OrderStatus.INVALID: "INVALID",
    OrderStatus.DENIED: "DENIED",
    OrderStatus.EMULATED: "EMULATED",
    OrderStatus.RELEASED: "RELEASED",
    OrderStatus.PENDING_UPDATE: "PENDING_UPDATE",
    OrderStatus.PENDING_CANCEL: "PENDING_CANCEL",
    OrderStatus.ACCEPTED: "ACCEPTED",
    OrderStatus.REJECTED: "REJECTED",
    OrderStatus.CANCELED: "CANCELED",
    OrderStatus.EXPIRED: "EXPIRED",
    OrderStatus.TRIGGERED: "TRIGGERED",
    OrderStatus.PENDING_NEW: "PENDING_NEW",
    OrderStatus.PARTIALLY_FILLED: "PARTIALLY_FILLED",
    OrderStatus.FILLED: "FILLED"
}

TIME_IN_FORCE_MAP = {
    TimeInForce.NO_TIME_IN_FORCE: "NO_TIME_IN_FORCE",
    TimeInForce.DAY: "DAY",
    TimeInForce.GTC: "GTC",
    TimeInForce.IOC: "IOC",
    TimeInForce.FOK: "FOK",
    TimeInForce.GTD: "GTD",
    TimeInForce.AT_THE_OPEN: "AT_THE_OPEN",
    TimeInForce.AT_THE_CLOSE: "AT_THE_CLOSE"
}

POSITION_SIDE_MAP = {
    PositionSide.NO_POSITION_SIDE: "NO_POSITION_SIDE",
    PositionSide.FLAT: "FLAT",
    PositionSide.LONG: "LONG",
    PositionSide.SHORT: "SHORT"
}

LIQUIDITY_SIDE_MAP = {
    LiquiditySide.NO_LIQUIDITY_SIDE: "NO_LIQUIDITY_SIDE",
    LiquiditySide.MAKER: "MAKER",
    LiquiditySide.TAKER: "TAKER"
}

File: crates/model/src/enums.rs
//! Core enumerations for AlphaForge trading system

use serde::{Serialize, Deserialize};

/// Order side enumeration
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[repr(u8)]
pub enum OrderSide {
    /// Buy order (bid)
    Buy = 1,
    /// Sell order (ask)  
    Sell = 2,
}

impl OrderSide {
    /// Get the opposite side
    pub fn opposite(self) -> Self {
        match self {
            Self::Buy => Self::Sell,
            Self::Sell => Self::Buy,
        }
    }
}

impl std::fmt::Display for OrderSide {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Buy => write!(f, "BUY"),
            Self::Sell => write!(f, "SELL"),
        }
    }
}

/// Order type enumeration
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[repr(u8)]
pub enum OrderType {
    /// Market order - executes at current market price
    Market = 1,
    /// Limit order - executes at specified price or better
    Limit = 2,
    /// Stop order - becomes market order when stop price hit
    Stop = 3,
    /// Stop-limit order - becomes limit order when stop price hit
    StopLimit = 4,
    /// Trailing stop order
    TrailingStop = 5,
}

impl std::fmt::Display for OrderType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Market => write!(f, "MARKET"),
            Self::Limit => write!(f, "LIMIT"),
            Self::Stop => write!(f, "STOP"),
            Self::StopLimit => write!(f, "STOP_LIMIT"),
            Self::TrailingStop => write!(f, "TRAILING_STOP"),
        }
    }
}

/// Order status enumeration
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[repr(u8)]
pub enum OrderStatus {
    /// Order is submitted but not yet accepted
    Submitted = 1,
    /// Order is accepted by the exchange
    Accepted = 2,
    /// Order is cancelled
    Cancelled = 3,
    /// Order is expired
    Expired = 4,
    /// Order is triggered (for stop orders)
    Triggered = 5,
    /// Order is partially filled
    PartiallyFilled = 6,
    /// Order is completely filled
    Filled = 7,
    /// Order is rejected
    Rejected = 8,
}

impl std::fmt::Display for OrderStatus {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Submitted => write!(f, "SUBMITTED"),
            Self::Accepted => write!(f, "ACCEPTED"),
            Self::Cancelled => write!(f, "CANCELLED"),
            Self::Expired => write!(f, "EXPIRED"),
            Self::Triggered => write!(f, "TRIGGERED"),
            Self::PartiallyFilled => write!(f, "PARTIALLY_FILLED"),
            Self::Filled => write!(f, "FILLED"),
            Self::Rejected => write!(f, "REJECTED"),
        }
    }
}

/// Time in force enumeration
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[repr(u8)]
pub enum TimeInForce {
    /// Good Till Cancelled
    GTC = 1,
    /// Immediate Or Cancel
    IOC = 2,
    /// Fill Or Kill
    FOK = 3,
    /// Good Till Date
    GTD = 4,
    /// Day order
    DAY = 5,
}

impl std::fmt::Display for TimeInForce {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::GTC => write!(f, "GTC"),
            Self::IOC => write!(f, "IOC"),
            Self::FOK => write!(f, "FOK"),
            Self::GTD => write!(f, "GTD"),
            Self::DAY => write!(f, "DAY"),
        }
    }
}

/// Book action for order book updates
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[repr(u8)]
pub enum BookAction {
    /// Add order to book
    Add = 1,
    /// Update existing order
    Update = 2,
    /// Delete order from book
    Delete = 3,
    /// Clear all orders at price level
    Clear = 4,
}

impl std::fmt::Display for BookAction {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Add => write!(f, "ADD"),
            Self::Update => write!(f, "UPDATE"),
            Self::Delete => write!(f, "DELETE"),
            Self::Clear => write!(f, "CLEAR"),
        }
    }
}

/// Aggressor side for trade ticks
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[repr(u8)]
pub enum AggressorSide {
    /// No aggressor (auction/crossing)
    NoAggressor = 0,
    /// Buyer was aggressor
    Buyer = 1,
    /// Seller was aggressor
    Seller = 2,
}

impl std::fmt::Display for AggressorSide {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::NoAggressor => write!(f, "NO_AGGRESSOR"),
            Self::Buyer => write!(f, "BUYER"),
            Self::Seller => write!(f, "SELLER"),
        }
    }
}

/// Instrument class enumeration
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[repr(u8)]
pub enum InstrumentClass {
    /// Spot instrument
    Spot = 1,
    /// Forward contract
    Forward = 2,
    /// Future contract
    Future = 3,
    /// Option contract
    Option = 4,
    /// Perpetual swap
    Perpetual = 5,
    /// Contract for difference
    CFD = 6,
    /// Bond
    Bond = 7,
    /// Cryptocurrency
    Crypto = 8,
}

impl std::fmt::Display for InstrumentClass {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Spot => write!(f, "SPOT"),
            Self::Forward => write!(f, "FORWARD"),
            Self::Future => write!(f, "FUTURE"),
            Self::Option => write!(f, "OPTION"),
            Self::Perpetual => write!(f, "PERPETUAL"),
            Self::CFD => write!(f, "CFD"),
            Self::Bond => write!(f, "BOND"),
            Self::Crypto => write!(f, "CRYPTO"),
        }
    }
}

/// Component state enumeration for lifecycle management
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[repr(u8)]
pub enum ComponentState {
    /// Component is initializing
    Initializing = 1,
    /// Component is initialized but not started
    Initialized = 2,
    /// Component is starting
    Starting = 3,
    /// Component is running
    Running = 4,
    /// Component is stopping
    Stopping = 5,
    /// Component is stopped
    Stopped = 6,
    /// Component is resuming from stopped state
    Resuming = 7,
    /// Component is in error state
    Error = 8,
    /// Component is disposed
    Disposed = 9,
}

impl std::fmt::Display for ComponentState {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Initializing => write!(f, "INITIALIZING"),
            Self::Initialized => write!(f, "INITIALIZED"),
            Self::Starting => write!(f, "STARTING"),
            Self::Running => write!(f, "RUNNING"),
            Self::Stopping => write!(f, "STOPPING"),
            Self::Stopped => write!(f, "STOPPED"),
            Self::Resuming => write!(f, "RESUMING"),
            Self::Error => write!(f, "ERROR"),
            Self::Disposed => write!(f, "DISPOSED"),
        }
    }
}
